Changed typings of updateQuery's previousQueryResult to be potentially undefined

Cellule opened the pull request:
Related to #12228

There's an outstanding issue where `previousQueryResult` sometimes, flakily, ends up `undefined`
This change aims to have the typings reflect the actual runtime behavior allowing devs to write safe code that can handle the missing `previousQueryResult`
An alternative could be to simply not call `mapFn` if `previousQueryResult` is missing. While neither is ideal nor fixes the underlying issue of "why" `previousQueryResult` is `undefined`. At least it would avoid bad crashes.

I've been rolling with this patch for years now and I haven't seen any bad side-effects so far. As far as I can tell, the case where `previousQueryResult` can be safely ignored and is likely going to be called again with actual data when it matters

### Typings Update for `updateQuery`:

* Changed typings of `updateQuery`'s `previousQueryResult` to be potentially undefined in `@apollo/client`.
* Updated `updateQuery` method in `src/core/ObservableQuery.ts` to reflect the new typings.
* Modified `UpdateQueryFn` type in `src/core/watchQueryOptions.ts` to allow `previousQueryResult` to be undefined.
* Updated `ObservableQueryFields` interface in `src/react/types/types.ts` to reflect the new typings.

### VSCode Configuration:

* Added `editor.codeActionsOnSave` setting to disable organizing imports on save in `.vscode/settings.json`.

<!--
  Thanks for filing a pull request on Apollo Client!

  A few automated bots may chime in on your PR. They are here to help
  with reviewing and ensuring Apollo Client is production ready after each
  pull request merge.

    - apollo-cla will respond asking you to sign the CLA if this is your first PR.
      It may also respond with warnings, messages, or fail the build if something is off.
      Don't worry, it'll help you to fix what is broken!

    - bundlesize is a status check to keep the footprint of Apollo Client as small as possible.

    - circleci will run tests, checking style of code, and generally make
      sure everything is working as expected

  Please look at the following checklist to ensure that your PR
  can be accepted quickly:

### Checklist:
- [X] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [X] Make sure all of the significant new logic is covered by tests
-->

jerelmiller commented on src/__tests__/subscribeToMore.ts:
```diff
@@ -241,6 +241,9 @@ describe("subscribeToMore", () => {
         }
       `,
       updateQuery: (prev, { subscriptionData }) => {
+        if (!prev) {
```
I definitely understand why you did this for this test, but unfortunately this makes it more difficult to know if we introduced regressions by accidentally making `prev` undefined when it should have a value. The assertion below isn't run in that case, so it would appear that the test would still pass.

Instead, I'd recommend updating the `prev` to `prev!`, that way this test will crash if this ever switches to `undefined`.

Cellule replied:
Make sense, I wasn't 100% sure how to handle it, but you're right the expectation is that `prev` should be provided here

jerelmiller reviewed (commented):
I'm generally ok with this change, but would you be willing to add a test that demonstrates the case where `previousData` is `undefined`? I'd like to make sure its "documented" in some form through runtime behavior, not just the types. That should help us prevent regressions in the future. That would also help determine whether this is an actual bug, or if this behavior was intentional. I'd like to avoid a band-aid to the types just in case `previousData` was never intended to be `undefined`.

Thanks for the contribution!

Cellule reviewed (commented):
> would you be willing to add a test that demonstrates the case where previousData is undefined

If you read the issue, the problem is that we can't figure out the repro steps to have `undefined` at runtime. 
It is likely a race condition that might be hard to reproduce in a testing environment.
I admit I haven't tried recently, this is an issue we've faced years ago and kept the patch. 
I can try to get a repro, but I can't make any guarantee

jerelmiller commented:
> If you read the issue, the problem is that we can't figure out the repro steps to have undefined at runtime.

Ah shoot. I read that and it went right over my head. Let me play around with this a bit more and see if I can think of any way this could happen. I'll push a test if I can, or respond back and let you know that I couldn't come up with something.

jerelmiller commented:
@Cellule the closest I can get to reproducing this is `updateQuery` passing an empty object `{}` as `previous`. Just to double check, the value you're seeing on the crashes is `undefined` and not access on a nested property when `previous` is `{}` correct? 

`updateQuery` calls `cache.diff` with `returnPartialData: true`: https://github.com/apollographql/apollo-client/blob/de8196eb72498e4e45438697ec17198e5b88bc44/src/core/ObservableQuery.ts#L732-L737

That `returnPartialData: true` returns an empty object if there is absolutely no cache data available for the given query. I've never actually seen `result` return `undefined` and have been unable to reproduce a situation where it returns `undefined` instead of `{}`. I've traced much of the code and am not seeing anything stick out either.

Is there anything else you can tell me about the queries where you see this happen, such as directives used, type policies, etc? Perhaps the combination of one of those things hits this case.

Cellule commented:
I admit it's been a long time since I've seen the issue because we've mitigated it by always checking `if(!prev) return prev!`
I am fairly certain `prev === undefined`
I'll try to get in a bad state on my side

jerelmiller commented:
That would be super helpful if you can. I fear this change is more of a band-aid to the real underlying issue and would love to figure that out if we can. Otherwise this could be a fairly disruptive change to a lot of existing users who could now potentially see a lot of TypeScript errors after upgrading.

Cellule commented:
Well maybe I was wrong, it seems it might be `{}` after all (so my patch wouldn't fix anything really)
<img width="1130" alt="image" src="https://github.com/user-attachments/assets/85711f83-487d-4c0b-a748-36234f5eb578" />

In my case I'm using `updateQuery` in `query.subscribeToMore` 
In order to repro I'm doing some operations that will cause the subscription to send more data then navigate to another page "quickly-ish"
The other page is still on the same entity, so it will effectively 
- change variables in `useQuery`
- Through `useEffect` it will resubscribe the subscription `
- `updateQuery` seems to be called with data from the previous variables for a new query with no data (hence empty object)

```json
{
    "subscriptionData": {
        "data": {
            "messageUpserted": {
                "message": {
                    "id": 9627,
                    "content": {
                        "text": "removed the work order from being on hold.",
                        "tokens": [
                            {
                                "value": "removed the work order from being on hold.",
                                "__typename": "TextToken"
                            }
                        ],
                        "__typename": "TokenizedText"
                    },
                    "author": {
                        "id": 1338,
                        "img": "https://img.maintainx-dev.com/mx-dev-cac1-uploads/static/user_placeholders/RandomPicture4.png",
                        "firstName": "Michael",
                        "lastName": "Ferris",
                        "displayName": "Michael Ferris",
                        "__typename": "User",
                        "alternateImg": null,
                        "availabilityStatus": null
                    },
                    "createdAt": "2025-01-22T19:50:41.676Z",
                    "reactions": [],
                    "__typename": "SystemWorkOrderStatusChangedMessage",
                    "extraData": {
                        "oldStatus": "ON_HOLD",
                        "newStatus": "OPEN",
                        "newStatusVariant": null,
                        "failures": null,
                        "escalatedUsers": [],
                        "escalatedTeams": [],
                        "__typename": "SystemWorkOrderStatusChangedExtraData"
                    }
                },
                "parent": {
                    "id": 876, // Note the id of the parent that the subscription subscribes to
                    "isUnreadForMe": false,
                    "comments": {
                        "totalCount": 0,
                        "unreadCount": 0,
                        "__typename": "MessageThread"
                    },
                    "__typename": "WorkOrder"
                },
                "transcription": null,
                "__typename": "MessageUpsertedSubscription"
            }
        }
    },
    "variables": {
        "noRedirect": false,
        "id": 4, // Note the id in the variables do not match with the data from the subscription.
        "pagination": {
            "cursor": "",
            "limit": 25
        },
        "shouldFetchAutomationInformation": false
    }
}
```

Cellule commented:
There's still something that bugs me with the `variables`
From what I can test, it's possible the variables passed to `subscribeToMore.updateQuery` do not match at all the variables passed to the subscription.
It seems entirely possible for the ObservableQuery to have receive new variables, but the subscription is still on the old variables

The scenario seems to be as follows
- Variables are changed (navigation or whatever)
- `useQuery` receives new variables and updates the underlying ObservableQuery
- `useEffect(() => query.subscribeToMore(...), [query.variables])` causes the subscription to be slated to be disconnected
- Data arrives in the subscription, the `updateQuery` is called for the previous variables, but the cache looks for query data using the new variables (leading to `{}`)
- `useEffect` unsubscribes the subscription then resubscribes correctly
It looks like a race condition, but so far it hasn't been hard to reproduce on my side

Real quick it sounds like it would be "safe" to simply not call `mapFn` in `updateQuery` if `results === {} || results === undefined`.
So far I haven't been able to reproduce the variables mismatch when there's a cache hit, but it sounds possible in theory

Another thing to mention, the `SubscribeToMoreOptions.updateQuery` implies the variables will have type `TSubscriptionVariables` but in reality, the variables returned are `TVariables` (not currently passed in the generics).
This type is simply wrong, while it's unlikely to be used, if the variables of the query and the variable of the subscription differ enough, it can cause bugs. I had to cast it just to confirm the bad state

```ts
makeSubscription<
            IWorkOrderDetailsQueryData,
            IWoMessageUpsertedSubscriptionSubscription,
            IWoMessageUpsertedSubscriptionSubscriptionVariables
          >("messageUpsertedSubscription", query.subscribeToMore, {
            document: messageUpsertedSubscription,
            variables: {
              parent: { id: workOrderId, type: IUserMessageParentType.WorkOrder },
            },
            updateQuery(prev, { subscriptionData, variables }) {
              const newMessage = subscriptionData?.data?.messageUpserted?.message;
              const workOrderUpdates = subscriptionData?.data?.messageUpserted?.parent;
              const transcription = subscriptionData.data.messageUpserted?.transcription;
              if (
                (variables as any as IWorkOrderDetailsQueryVariables).id !==
                (workOrderUpdates as any).id
              ) {
                debugger;
                return prev;
              }
              if (!newMessage || workOrderUpdates?.__typename !== "WorkOrder" || !prev?.workOrder) {
                return prev!;
              }
```

Cellule commented:
One issue I see with using `DeepPartial` is that it will then become a typings nightmare to properly return the correct object

<img width="1042" alt="image" src="https://github.com/user-attachments/assets/f17220a6-8789-4c74-8f35-b726b2c51f27" />

I wonder if a better alternative is to actually only update the query if it is complete
I had that idea in the beginning because in my opinion, if for whatever reason the previous data is missing or incomplete, I will want to no-op the update.
It does make a difference in the runtime and could potentially affect some legit use case (although I can't think of one) I feel it might be the better/safer approach.
Likely, if you have a good reason to want to write in the cache of an incomplete query, you're more likely to call `cache.writeQuery` manually than rely on `updateQuery`

```ts
public updateQuery<TVars extends OperationVariables = TVariables>(
    mapFn: (
      previousQueryResult: Unmasked<TData>,
      options: Pick<WatchQueryOptions<TVars, TData>, "variables">
    ) => Unmasked<TData> | undefined
  ): void {
    const { queryManager } = this;
    const { result, complete } = queryManager.cache.diff<Unmasked<TData>>({
      query: this.options.query,
      variables: this.variables,
      returnPartialData: true,
      optimistic: false,
    });

    if (complete && result) {
      const newResult = mapFn(result, {
        variables: (this as any).variables,
      });
      if (newResult) {
        queryManager.cache.writeQuery({
          query: this.options.query,
          data: newResult,
          variables: this.variables,
        });

        queryManager.broadcastQueries();
      }
    }
  }
```

jerelmiller commented:
I go back and forth on this. This sounds like a reasonable change, but I fear its too much of a breaking change for a patch/minor release. Since we don't provide any additional information in the callback (i.e. whether the previous result is complete), there could be cases where suddenly your callback is no longer called and its difficult to tell why.

That said, the callback does take an `options` argument as the 2nd argument. Perhaps we can provide the `complete` flag here and let the user decide whether to use the partial data to return a full result or to ignore the update (i.e. return `undefined`). At least this gives the best of both worlds and puts it into the user's hands on how to handle it.

Looking at the return type of `updateQuery`, currently its set to `Unmasked<TData>`, but it does appear the implementation is [ok returning a falsey value](https://github.com/apollographql/apollo-client/blob/ce6e37f61dc1037b20972bd8eba31434cba356ce/src/core/ObservableQuery.ts#L743-L751) and will do nothing if thats the case. We should probably reflect this in the types as well by updating the return type to `Unmasked<TData> | undefined` so that TypeScript allows this.

The tricky part would be updating the TypeScript type here. It would be ideal if it were something like this:

```ts
updateQuery(previous, { complete }) {
  if (complete) {
    previous;
  // ^? TData
  }

  previous;
  // ^? DeepPartial<TData>
}
```

Thoughts on this?

Cellule commented:
> I go back and forth on this. This sounds like a reasonable change, but I fear its too much of a breaking change for a patch/minor release. Since we don't provide any additional information in the callback (i.e. whether the previous result is complete), there could be cases where suddenly your callback is no longer called and its difficult to tell why.
> 

Yeah this is exactly why I'm on the fence with the change as well

> That said, the callback does take an `options` argument as the 2nd argument. Perhaps we can provide the `complete` flag here and let the user decide whether to use the partial data to return a full result or to ignore the update (i.e. return `undefined`). At least this gives the best of both worlds and puts it into the user's hands on how to handle it.
> 

Humm yeah that could work, but indeed the typings would be the tricky part

> Looking at the return type of `updateQuery`, currently its set to `Unmasked<TData>`, but it does appear the implementation is [ok returning a falsey value](https://github.com/apollographql/apollo-client/blob/ce6e37f61dc1037b20972bd8eba31434cba356ce/src/core/ObservableQuery.ts#L743-L751) and will do nothing if thats the case. We should probably reflect this in the types as well by updating the return type to `Unmasked<TData> | undefined` so that TypeScript allows this.
> 

Right, I did want to update at least the return type which is more accurate and backward compatible.
This is why I have to do this in my code currently `updateQuery(prev) { if (!prev) return prev!; ... }`

> The tricky part would be updating the TypeScript type here. It would be ideal if it were something like this:
> 
> ```ts
> updateQuery(previous, { complete }) {
>   if (complete) {
>     previous;
>   // ^? TData
>   }
> 
>   previous;
>   // ^? DeepPartial<TData>
> }
> ```
> 
> Thoughts on this?

Keeping backward compatibility and typescript relation between the 2 variables sounds like a challenge!
I'll try to give it some thoughts, but whatever solution it is likely to cause friction after the update, which I think is unavoidable

jerelmiller commented:
> whatever solution it is likely to cause friction after the update, which I think is unavoidable

Definitely unfortunate, but I'd wager its more preferable than a crash in production!

Cellule commented:
Alright here's my attempt

This shows my reasoning best I think 

```ts
const updateQuery: Parameters<typeof observable.updateQuery>[0] = jest.fn(
      (previousResult) => {
        expect(previousResult.complete).toBe(true);
        // Type guard
        if (!previousResult.complete) {
          return undefined;
        }
        return { user: { ...previousResult.user, name: "User (updated)" } };
      }
    );
```

There's no way I can think of to bind the type of a parameter with a type guard on another parameter.
The details can be tweaked, but my idea is to attach a `complete` property straight on the result. Obviously this doesn't work as-is since the result could have a `complete` property.
In the callback, if you don't check for `prev.complete` the type will be `Partial<Unmasked<TData>>` (I gave up on `DeepPartial` because it was giving me typescript errors due to type being too complex)

I cleaned up a bunch of types that we're not shared where it should
I fixed the typings for `variables` in the options to be `TVariables` and added a new `subscriptionVariables: TSubscriptionVariables` in the subscribeToMore case

Let me know what you think

jerelmiller commented:
This is turning out to be a much more difficult change than expected 🙃.

Unfortunately adding a `complete` field to the result itself is a no-go. We want to avoid doing anything with `data` since it could conflict with the fields returned from the GraphQL result. We do our best to avoid that at all costs.

Unfortunately changing that first argument in any capacity is a breaking change. That said, how urgent is this change for you? We have started working on v4 and we could certainly think about moving this change there. I'd be open to redesigning the `updateQuery` API to be more type-safe and v4 would give us a chance to make any necessary breaking changes to facilitate this.

Thoughts?

Cellule commented:
> This is turning out to be a much more difficult change than expected 🙃.
> 
> Unfortunately adding a `complete` field to the result itself is a no-go. We want to avoid doing anything with `data` since it could conflict with the fields returned from the GraphQL result. We do our best to avoid that at all costs.
> 
> Unfortunately changing that first argument in any capacity is a breaking change. That said, how urgent is this change for you? We have started working on v4 and we could certainly think about moving this change there. I'd be open to redesigning the `updateQuery` API to be more type-safe and v4 would give us a chance to make any necessary breaking changes to facilitate this.
> 
> Thoughts?

I understand that changing the first argument doesn't work, I went down that route to see how feasible it would be.
Since we're considering the v4 path, one change I'm thinking that would make sense is to provide an option in updateQuery to skip the update on partial results.
For 3.13 we could default that option to always run even on partial data, but on v4 we could flip the default so we'd always skip on partial data unless explicitly requested by the user.
That way people who care about it now can provide the option to skip it and in the future the behavior will make more sense.
This way we don't touch the typings, it won't be perfect, but it at least provides a runtime way to avoid the crash

jerelmiller commented:
> one change I'm thinking that would make sense is to provide an option in updateQuery to skip the update on partial results

This is technically already possible if your callback function doesn't return a value:
https://github.com/apollographql/apollo-client/blob/eeba2efe1df86e770807d3737b256f59f4d5690e/src/core/ObservableQuery.ts#L743-L751

If v4 gives you the right types/values to determine when you have an incomplete result, this would just become:

```ts
// pseudo-code for what v4 might look like API-wise
observableQuery.updateQuery(({ previousData, complete }) => {
  if (!complete) {
    return;
  }
});
```

I think we could probably just stick with this. It would simplify the types as well since we wouldn't need to account for that option in order to type the arguments correctly (i.e. you wouldn't need to handle `skipForPartialData: true` and handle typing `data` differently for `true` vs `false`).

> For 3.13 we could default that option to always run even on partial data

The problem here is that it doesn't solve for the type issue which is the biggest problem IMO with the existing implementation. Even if we add that option for you to toggle off in 3.13, the types still wouldn't be correct for the case when the function is called with partial data.

That said, something that does seem to work is if `data` is part of the same object as that `complete` flag. Perhaps we consider deprecating the first argument and ask devs to read from the 2nd one if they need that type safety:

```ts
observableQuery.updateQuery(_, { previousData, complete }) => {
  if (complete) {
    // previousData is TData
  }

  if (!complete) {
    // previousData is DeepPartial<TData>
  }

  // previousData is TData | DeepPartial<TData>
})
```

This gives us an on-ramp to 4.0 which would just remove that first argument.

Certainly not the most elegant for 3.13, but it would get us to a type-safe solution now. Check out [this playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3nAqgOxAQwGcBrFAEwBo4B5VKfGYCHANXymHwCMAbFQuAF84AMygQQcAEQABfJB48IAegDGPYChwwpAbgBQoSLATI0mACIoUYAArtG+HkNHjJs+REUr1m7coBXRg1Gfj19fXUiAWouQhQoADduPgBFAITEAB59ODgAFQsGfDgAXjh8HEQKXIK2DhT+OBQADxgtMhi6BiZWdk5eJvLaBJ7meoG+QhqAPgxasADeYFU4ALAyBhR0zIAKWryCMAAxHAAuOH2867gAOnv2AHNCM4ObuAAfOABtN-frsBQFCJJgBQg7KCIABK-ACPBgF1wBBI5CyhWKMxq-3+EDAjGYL3m2OxyQag0J+QmjUIBmJ-1UEjAfHaFxgUAytLpN0BwNBhCKMHwiLwRFIZDRAvwM05XMEWLpAF0-tcvr8uXkeSCIGCIdDYfCLlYbPZYJweFkkaLUejBTNMcqcXjeoT0A6Sf1qRdKR7yTL1QzwMyUBcRE54n6uZq+ZLDdY7A4zRaRSjxTapdK3Tc5Zm4EqbgBKMpzS0piUYz5wRIQYCUWr5i5VmsYQT6FuRAnwCBxBLJQZlOA4FAAdxo3aSjV1WUwIggEFZ7LQghmu3zBn0Xfi48Gt3Wm3aut2uyj2vBGUhMMIcJgVEwAaZKHaVGPYMlQkLpTmrrywBElzvQcLL9rmUZRsBwGdYACHAth4RACjMABlVQODxOAyAgfgcAAcngaCoHEEcYAAC2AAQkFQOAGXwlBVBgWDbgKEiBFIt4QLgIiEjQGAIDgQgiIgQiOLQn8RE4nBVDQLgHyHawcAQISYCHCBbjeZ9T0yC8r1Y0CAD0AH5VKBLUX2KP02P02o22-X9dgAQn-B8UEAwzeRPXVNPhbS4As7kjOjYovJ8oQIg1Py3LPPVL08vJzIM0LXJMwValikK4CBGAAigHBW1XfQgA) which shows this working.

jerelmiller commented on src/core/watchQueryOptions.ts:
```diff
@@ -164,31 +164,98 @@ export interface FetchMoreQueryOptions<TVariables, TData = any> {
+     * @deprecated Use `options.previousQueryResult` instead.
+     */
+    previousQueryResult: Unmasked<TData>,
+    options: TOptions & UpdateQueryOptions<TData, TVariables>
```
Unfortunately I don't think we can use the `{}` here for `TOptions` so I'd like to see if we can come up with a different solution.

I recently discovered that the `@typescript-eslint/no-unsafe-*` rules complain when you enable [typed linting](https://typescript-eslint.io/getting-started/typed-linting) in ESLint and using the `{}` violates these rules. I installed this PR build in the [spotify showcase](https://github.com/apollographql/spotify-showcase) which has these rules enabled and I see ESLint warnings (which would be errors if I set the severity as such):
<img width="837" alt="Screenshot 2025-02-05 at 2 04 16 PM" src="https://github.com/user-attachments/assets/9378fbc2-5c86-4ee4-ba8f-88b08698a23a" />

I definitely understand the motivation behind the `TOptions` type, but I think we should probably create separate `*UpdateQueryFn` types for each unique callback type which provides their respective options.

jerelmiller commented on src/core/watchQueryOptions.ts:
```diff
@@ -164,31 +164,98 @@ export interface FetchMoreQueryOptions<TVariables, TData = any> {
+  TOptions = {},
+> {
+  (
+    /**
+     * @deprecated Use `options.previousQueryResult` instead.
+     */
```
Can we be a bit more verbose here? I think having a bit of a description on why this is an issue will help users understand why these need to stop using this argument. Feel free to workshop this suggestion:

```suggestion
    /**
     * @deprecated This value is not type-safe and may contain partial data. This
     * argument will be removed in the next major version of Apollo Client. Use
     * `options.previousData` instead for a more type-safe value.
     */
```

jerelmiller commented on src/react/hoc/__tests__/queries/updateQuery.test.tsx:
```diff
@@ -41,6 +41,7 @@ describe("[queries] updateQuery", () => {
           try {
             data!.updateQuery(() => {
               done = true;
+              return undefined;
```
See my other comment in regards to changing the return type to `void` which will avoid the need for this change.

jerelmiller commented on src/react/hooks/__tests__/useBackgroundQuery.test.tsx:
```diff
@@ -8411,16 +8453,51 @@ describe.skip("type tests", () => {
+            complete,
+            previousQueryResult,
+          }
+        ): UnmaskedVariablesCaseData => {
```
I'm curious, is this needed? This might indicate a code smell if the type of `updateQuery` doesn't use the unmasked type. If this is working without the type, I'd remove this and let the type of `updateQuery` dictate this:

```suggestion
        ) => {
```

jerelmiller reviewed (changes requested):
Ok this should be my last round of feedback. This is looking great though otherwise!

Let me know if you'd like me to make the updates and I'd be happy to push the changes so we can get this across the finish line. I'm planning to release a 3.13 beta in the next day or two and would love to have this in there 🙂

Cellule commented on src/core/watchQueryOptions.ts:
```diff
@@ -164,31 +164,98 @@ export interface FetchMoreQueryOptions<TVariables, TData = any> {
 }
 
-export type UpdateQueryFn<
+export interface UpdateQueryFn<
```
I'm confused by my changes at this point, will push the other fixes and review this

Cellule replied:
I checked again and I'm not 100% sure anymore. Public types are types exported from src/core/index.ts right? If so then the type we need to be backward compatible is `UpdateQueryOptions` not `UpdateQueryFn` 
I essentially added `complete` and `previousData` to `UpdateQueryOptions` which I think is backward compatible right?
What are your rules when it comes to exporting types publicly? 
I think I added a few, maybe too many?

Cellule commented on src/react/hooks/useLoadableQuery.ts:
```diff
@@ -269,7 +269,10 @@ export function useLoadableQuery<
-        return internalQueryRef.observable.subscribeToMore(options);
+        return internalQueryRef.observable.subscribeToMore(
+          // TODO: The internalQueryRef doesn't have TVariables' type information so we have to cast it here
+          options as any as SubscribeToMoreOptions<TData, OperationVariables>
```
I think you are right about the other call sites, but here there seems to be an incompatibility due to TSubscriptionData vs TData

```
Conversion of type 'SubscribeToMoreOptions<TData, TSubscriptionVariables, TSubscriptionData, TVariables>' to type 'SubscribeToMoreOptions<TData, OperationVariables, TData, OperationVariables>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'TSubscriptionData' is not comparable to type 'TData'.
    'TData' could be instantiated with an arbitrary type which could be unrelated to 'TSubscriptionData'.ts(2352)
```

jerelmiller commented on src/core/ObservableQuery.ts:
```diff
@@ -722,23 +723,32 @@ Did you mean to call refetch(variables) instead of refetch({ variables })?`,
+  public updateQuery(mapFn: UpdateQueryMapFn<TData, TVariables>): void {
     const { queryManager } = this;
-    const { result } = queryManager.cache.diff<TData>({
+    const { result, complete } = queryManager.cache.diff<Unmasked<TData>>({
```
Let's add a `TODO` comment then so I don't forget to remove this when I get around to fixing that type if thats ok with you.

Cellule replied:
Ah ok I understand better your point, you mean `cache.diff<TData>` should likely return type `Unmasked<TData>` for result. Ok fair I'll revert

jerelmiller commented on src/core/ObservableQuery.ts:
```diff
@@ -538,6 +537,7 @@ Did you mean to call refetch(variables) instead of refetch({ variables })?`,
                     optimistic: false,
                   },
                   (previous) =>
+                    // REVIEW: Code smell here with the `!` and cast, is it possible for previous to be null or have partial data ?
```
Let's address separately since it was already there. Feel free to remove this for now. `cache.updateQuery` might be a whole other animal.

Cellule commented:
Alright I think it mostly looks good.
If you think there are other things that needs to be fixed I won't mind if you decide to take care of it yourself. I will likely not have more time before next week as I'm working on a presentation for Fragment Masking to my company this Friday :)

jerelmiller commented:
Sounds great! Really appreciate the back and forth. I'll make sure this change makes it into the 3.13 beta 🙂.

Hope the presentation goes well! Feel free to ping me on Discord (`@jerelmiller`) if you've got questions about data masking 🙂. I'd love to answer them!
